juspay/hyperswitch
# feat(webhooks): store request and response payloads in `events` table

## Type of Change



## Description

This PR includes the following changes:

- Adds a few columns to the `events` table that allow for storing request and response headers and payloads.
- Introduces domain types for events.
- Stores the request headers and payloads, and the response headers, payloads and status codes in the `events` table.
- Updates the consumer workflow to use the previously stored request information (instead of retrieving the current resource information) when retrying deliveries of failed webhooks.
  - This includes backward-compatible changes for tasks added by previous versions of the `router` service, thus allowing for staggered deployments of the `router` service.
  - For events added by the previous version of the `router` service, the workflow fetches the resource information from the database and constructs the webhook request (as in the version of the `router` service without these changes).

**This PR introduces the following behavior changes:**
 
- Each webhook delivery retry attempt creates a new record in the `events` table.
- Webhook deliveries are retried irrespective of whether the resource transitions to a new state or not (the current status of the resource is not even checked when retrying deliveries).
- Each webhook delivery attempt has the exact same payload and headers, thus the signature should be the exact same.

_Note to reviewers: This PR can be reviewed one commit at a time, if you'd prefer._

### Additional Changes




The migrations for this PR are available in [`up.sql`](https://github.com/juspay/hyperswitch/blob/90975afa40eb999733a48291e79a569d41afb5cf/migrations/2024-03-04-204051_events_store_webhook_delivery_attempt_info/up.sql) and [`down.sql`](https://github.com/juspay/hyperswitch/blob/90975afa40eb999733a48291e79a569d41afb5cf/migrations/2024-03-04-204051_events_store_webhook_delivery_attempt_info/down.sql) files.

**Note that since the migrations included drop a couple of columns, the destructive queries can be deferred for later: the application will run fine even if the extra columns exist.**

## Motivation and Context


This would allow for better handling of cases where the resource transitions to a new state after the event was created. This change also paves the way for manually retrying the delivery of a specific webhook.

## How did you test it?


Firstly, I've configured the retries to happen at shorter intervals than the default configuration:

```shell
curl --location 'http://localhost:8080/configs/' \
  --header 'Content-Type: application/json' \
  --header 'Accept: application/json' \
  --header 'api-key: <ADMIN_API_KEY>' \
  --data '{
    "key": "pt_mapping_outgoing_webhooks",
    "value": "{\"default_mapping\":{\"start_after\":60,\"frequency\":[15,30],\"count\":[2,3]},\"custom_merchant_mapping\":{}}"
  }'
```

As per this configuration, the maximum number of retries is 7 (1 initial attempt, 1 after `start_after`, 2 + 3 at specified frequencies).

1. **Each webhook delivery attempt creates new entries in the `events` table:**

   1.  This can be verified using a query like so:
       ```sql
       SELECT event_id, event_type, event_class, primary_object_id, primary_object_type, created_at, primary_object_created_at, initial_attempt_id, idempotent_event_id, is_webhook_notified FROM events WHERE primary_object_id = 'pay_9fR0FvtmjKNxaYFhKXCY' AND event_type = 'payment_succeeded' ORDER BY created_at DESC;
       ```

       ![Screenshot of delivery attempts in `events` table](https://github.com/juspay/hyperswitch/assets/22217505/cbadaa15-2694-4a85-8d4d-ce24a70bd41f)

   2. Note that:
      1. Each `event_id` is unique, and is time-sorted (`evt` prefix + UUID v7).
      2. `idempotent_event_id` follows the old format for the initial attempt, while including a random suffix for retries.
      3. Other details like `event_type`, `event_class`, `primary_object_id`, `primary_object_type`, `primary_object_created_at` and `initial_attempt_id` are the same for all records.
      4. The values stored in the `request` and `response` fields are not human-readable since they're encrypted. However, we can partially verify that the `request` field contains the same data by verifying the equality of the webhook headers and payloads sent in each webhook delivery attempt. The values stored in the `response` field can be verified once events retrieve APIs are available.

2. **Webhook delivery retries happen irrespective of the resource transitioning to a different state:**

   1. Create a payment with a Stripe disputed test card (say `4000000000002685`) which sends two webhooks: one each for `dispute_opened` and `dispute_lost`.
   2. Verify that each event is retried independently:

      ```sql
      SELECT event_id, event_type, event_class, primary_object_id, primary_object_type, created_at, primary_object_created_at, initial_attempt_id, idempotent_event_id, is_webhook_notified FROM events WHERE primary_object_id = 'dp_rgQkosgs0GTkKlK51210' ORDER BY created_at DESC;
      ```
      ![Screenshot of multiple webhooks related to same resource being retried independently](https://github.com/juspay/hyperswitch/assets/22217505/cab4634b-48e4-4cb1-9719-c2bfc030ed10)

3. Ensure that each retry attempt sends the exact same headers and payload:

   1. <details><summary>Older attempt (marked with orange circle) (click to view screenshot):</summary><img alt="Screenshot of attempt 1" src="https://github.com/juspay/hyperswitch/assets/22217505/de36e58a-37a6-4b0e-a927-c6c015095c67"></details>

   2. <details><summary>Newer attempt (marked with yellow circle) (click to view screenshot):</summary><img alt="Screenshot of attempt 2" src="https://github.com/juspay/hyperswitch/assets/22217505/3773998b-c246-4eb1-937b-c0f3adfd0a57"></details>

   3. <details><summary>Ensure that there is no diff in the payloads (click to view screenshot) (I used <a href="https://quick-diff.com">this online tool</a>):</summary><img alt="Screenshot of diff" src="https://github.com/juspay/hyperswitch/assets/22217505/2a32c43f-9fce-4827-8ac4-a06ef1f254b6"></details>

   5. <details><summary>Optionally, verify that the webhook signature is the same in both cases (click to view screenshot) (I used <a href="https://iotools.cloud/tool/hmac-generator">this online tool</a>):</summary><img alt="Screenshot of signature verification" src="https://github.com/juspay/hyperswitch/assets/22217505/7b302d37-92a8-404d-8bd9-46a2c0b0a091"></details>


## Checklist
